contentcuration/contentcuration/static/js/edit_channel/sharedComponents/Uploader.vue

```diff
@@ -0,0 +1,276 @@
+        type: Boolean,
+        default: false,
+      },
+      allowDrop: {
+        type: Boolean,
+        default: true,
+      },
+    },
+    data() {
+      return {
+        highlight: false,
+        unsupportedFiles: [],
+        tooLargeFiles: [],
+        totalUploadSize: 0,
+      };
+    },
+    computed: {
+      ...mapGetters('fileUploads', ['getFile']),
+      acceptedFiles() {
+        let filter = { display: true };
+        if (this.presetID) {
+          filter.id = this.presetID;
+        } else {
+          filter.supplementary = false;
+        }
+        return _.where(Constants.FormatPresets, filter);
+      },
+      acceptedMimetypes() {
+        return _.chain(this.acceptedFiles)
+          .pluck('associated_mimetypes')
+          .flatten()
+          .value()
+          .join(',');
+      },
+      acceptedExtensions() {
+        return _.chain(this.acceptedFiles)
+          .pluck('allowed_formats')
+          .flatten()
+          .uniq()
+          .value();
+      },
+      unsupportedFilesText() {
+        return this.$tr('unsupportedFilesText', {
+          count: this.unsupportedFiles.length,
+          extensions: this.acceptedExtensions.join(this.$tr('listDelimiter')),
+          extensionCount: this.acceptedExtensions.length,
+        });
+      },
+      availableSpace() {
+        return State.current_user.get('available_space');
+      },
+      highlightDropzone() {
+        return this.highlight && !this.readonly && this.allowDrop;
+      },
+      maxFileSize() {
+        return MAX_FILE_SIZE;
+      },
+    },
+    methods: {
+      // Add in once global store is properly set up
+      ...mapActions('fileUploads', ['uploadFile']),
+      ...mapMutations('fileUploads', {
+        addFile: 'ADD_FILE',
+        setFileError: 'SET_FILE_ERROR',
+      }),
+      enter() {
+        this.highlight = true;
+      },
+      over() {
+        this.highlight = true;
+      },
+      leave() {
+        this.highlight = false;
+      },
+      drop(e) {
+        this.highlight = false;
+        if (this.allowDrop) this.handleFiles(e.dataTransfer.files);
+      },
+      openFileDialog() {
+        if (!this.readonly) {
+          this.$refs.fileUpload.click();
+        }
+      },
+      setError(fileID, errorType) {
+        let message;
+        switch (errorType) {
+          case fileErrors.TOO_LARGE:
+            message = this.$tr(fileErrors.TOO_LARGE, { size: this.formatFileSize(MAX_FILE_SIZE) });
+            break;
+          case fileErrors.WRONG_TYPE:
+            message = this.$tr(fileErrors.WRONG_TYPE, {
+              filetypes: this.acceptedExtensions.join(', '),
+            });
+            break;
+          case fileErrors.NO_STORAGE:
+            message = this.$tr(fileErrors.NO_STORAGE);
+            break;
+          default:
+            message = this.$tr(fileErrors.UPLOAD_FAILED);
+        }
+        this.setFileError({ id: fileID, error: errorType, message: message });
+      },
```

Comment:
I believe that switch statements are often more cleanly written as if/else statements in javascript, because they have arcane and error-prone syntax.

I would also recommend _not_ re-using the external constants as i18n identifiers, because it will make it harder to find the relationship between strings uploaded to Crowdin and their usage in the code-base.

```javascript
setError(id, error) {
  let message = this.$tr('uploadFailedError');
  if (error === fileErrors.TOO_LARGE) {
    message = this.$tr('tooLargeError', { size: this.formatFileSize(MAX_FILE_SIZE) });
  } else if (error === fileErrors.WRONG_TYPE) {
    message = this.$tr('wrongTypeError', { filetypes: this.acceptedExtensions.join(', ') });
  } else if (error === fileErrors.NO_STORAGE) {
    message = this.$tr('noStorageError');
  }
  this.setFileError({ id, error, message });
}
```